File: answers/09-pre/main.rb
require 'pry'

def sixes(n)
  tmp = '6' * n
  result = []
  n.times { |i|
    result << tmp.dup
    tmp[i] = '9'
  }
  result << tmp
  result
end

def solve(input)
  nth, cards = input.split(':')
  result = sixes(cards.count('6')).flat_map {|six|
    pats = (cards.delete('6') + six).chars.permutation(4).to_a
    pats = pats.reject {|s| s[0] == '0'}
  }.sort.uniq[nth.to_i-1]
  result ? result.join : "-"
end

binding.pry

File: answers/09-pre/Main.hs
module Main where

import Data.List
import Data.List.Split

solve :: String -> String
solve input = if n < length patterns then patterns !! n else "-"
  where
    [nth, cards] = splitOn ":" input
    n = read nth - 1
    patterns = enumerate cards

enumerate :: String -> [String]
enumerate = nub . sort . filterCards . map (take 4) . concatMap permutations . check6

filterCards :: [String] -> [String]
filterCards = filter ((/= '0') . head)

check6 :: String -> [String]
check6 ns = map (rest ++) result
  where
    (six,rest) = partition (=='6') ns
    rep = tails . replicate (length six)
    result = zipWith (++) (rep '6') (reverse $ rep '9')
